students/barecki_andrzej/lesson_11_argparse_and_math/filling_in_the_gaps.py

```diff
@@ -0,0 +1,102 @@
+import os
+import datetime
+import re
+import argparse
+import sys
+
+
+def create_example_test_files(path, source_folder_param):
+    os.chdir(path)
+    print('Current working directory:', os.getcwd() + os.linesep)
+    src_dir = ""
+
+    """Create source folder(extend date/time approach)"""
+    if source_folder_param is not None:
+        if os.path.exists(source_folder_param):
+            random_folder = os.path.join(source_folder_param + '_' +
+                                         str(datetime.datetime.now().date()) +
+                                         '_' + str(datetime.datetime.now().
+                                                   time()).replace(':', '.'))
+            src_dir = random_folder
+        else:
+            src_dir = source_folder_param
+
+    """ Create example of few files """
+    file_list = [os.path.join(src_dir, "spam002.txt"),
+                 os.path.join(src_dir, "spam003.tyt"),
+                 os.path.join(src_dir, "spam004.txt"),
+                 os.path.join(src_dir, "spam005.jpg"),
+                 os.path.join(src_dir, "saam006.txt"),
+                 os.path.join(src_dir, "saam007.mp3"),
+                 os.path.join(src_dir, "spam007.txt"),
+                 os.path.join(src_dir, "spam009.txt"),
+                 ]
+
+    for elem in file_list:
+        if not os.path.exists(os.path.dirname(elem)):
+            os.makedirs(os.path.dirname(elem))
+        with open(elem, "w") as test_file:
+            test_file.close()
+
+    return os.path.join(os.getcwd(), src_dir)
+
+
+def filling_in_the_gap(path, prefix, extension):
+    spam_regex = re.compile(r'({0})(\d\d\d)({1})'.format(prefix, extension))
+    init = False
+    index = 0
+
+    """search all files from a given directory"""
+    for folder_name, _, file_names in os.walk(path):
+        for filename in sorted(file_names):
+            result = spam_regex.search(filename)
+            if result is not None:
+                """index initialization """
+                if init is False:
+                    index = int(result.group(2))
+                    init = True
+                    print("Initial file:{0}".format(os.path.join(folder_name,
+                                                                 filename)))
+                    print("Initial index:{0}".format(index) + os.linesep)
+                else:
+                    """ Update index """
+                    index = index + 1
+                    old_path = os.path.join(folder_name, filename)
+
+                    """Generate filename"""
+                    new_filename = "spam" + str(index).zfill(3) + ".txt"
+                    new_path = os.path.join(folder_name, new_filename)
+                    print("Before:{0}".format(old_path))
+
+                    """Update filename"""
+                    os.rename(old_path, new_path)
+                    print("After: {0}".format(new_path))
+
+
+def check_arg(args=None):
```

Comment:
1. Bardziej odpowiednia byłaby nazwa `parse_args`
2. Nie rozumiem po co jest domyśla wartość argumentu args. Chyba nie ma sytuacji w jakiej chcielibyśmy to wywołać bez listy argumentów

Reply:
@AndrzejBarecki zgadza się, unit testy byłyby tu odpowiednie.
Domyślą wartość argumentów stosuje się przeważnie dla danych konfiguracyjnych kiedy dla większości wywołań funkcja ma działać w jakiś sposób, ale chcemy zostawić pole do konfiguracji, np. open https://docs.python.org/3/library/functions.html#open 
Pierwszy argument jest wymagany, bo z załozenia zawsze chcemy jakis konkretny plik. Pozostałe argumenty są to opcjonalne dane konfiguracyjne z dostarczonymi domyślnymi wartościami, które są odpowiednie dla większości użyć tej funkcji.